core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/ReadOnlyLDAPUserStoreManager.java

```diff
@@ -1196,7 +1205,9 @@ private boolean bindAsUser(String userName, String dn, Object credentials) throw
                 log.debug("Authentication failed " + e);
                 log.debug("Clearing cache for DN: " + dn);
             }
-            userCache.remove(userName);
+            if (userName != null) {
+                userCache.remove(userName);
+            }
```

Comment:
If username is not significant in this function, cannot we simply throw the exception here and in the catch of the caller fucntion, remove username from cache ? Since this is a private function, you can consider doing that change very easily.